polkadot/runtime/parachains/src/assigner_on_demand/mod.rs

```diff
@@ -780,6 +828,53 @@ where
 		})
 	}
 
+	/// Provide the amount of revenue accumulated from Instantaneous Coretime Sales from Relay-chain
+	/// block number last_until to until, not including until itself. last_until is defined as being
+	/// the until argument of the last notify_revenue message sent, or zero for the first call. If
+	/// revenue is None, this indicates that the information is no longer available. This explicitly
+	/// disregards the possibility of multiple parachains requesting and being notified of revenue
+	/// information.
+	///
+	/// The Relay-chain must be configured to ensure that only a single revenue information
+	/// destination exists.
+	pub fn notify_revenue(when: BlockNumberFor<T>) -> DispatchResult {
+		let now = <frame_system::Pallet<T>>::block_number();
+		// When cannot be in the future.
+		ensure!(when <= now, Error::<T>::RequestedFutureRevenue);
+
+		// TODO actual revenue
+		let revenue: Balance = 0u128;
+		let message = Xcm(vec![
+			Instruction::UnpaidExecution {
+				weight_limit: WeightLimit::Unlimited,
+				check_origin: None,
+			},
+			mk_coretime_call(CoretimeCalls::NotifyRevenue(revenue)),
+		]);
+		if let Err(err) = send_xcm::<T::SendXcm>(
+			Location::new(0, [Junction::Parachain(T::BrokerId::get())]),
+			message,
+		) {
+			log::error!("Sending `NotifyCoreCount` to coretime chain failed: {:?}", err);
+		}
+
+		Ok(())
+	}
+
+	fn get_revenue(now: BlockNumberFor<T>, when: BlockNumberFor<T>) -> Balance {
+		let revenue: Balance = 0u128;
+		T::Revenue::mutate(|revenue| {
+			revenue.into_iter().enumerate().for_each(|(index, block_revenue)| {
+				//
+				if when >= now.saturating_sub(index) {
+					revenue.saturating_add(block_revenue);
+					block_revenue = 0;
```

Comment:
Well, that *might* make some sense. But I'm not really sure what is more optimal from the storage perspective: keeping a constant-length bounded vector or squeezing it for short periods of time when we don't need the tail elements. Trying to win on space here doesn't make sense, as numbers are too low, and it will be filled again during the next 160 blocks at max. And I'm afraid that constantly changing the length would make the vector jump between storage buckets. @kianenigma do you maybe have some insight?

Reply:
Left an alternative implementation in comments: https://github.com/paritytech/polkadot-sdk/pull/3940/files#diff-96442c27c75e4ced42fe956c99adf0b5651f35865fef7f42ebe2791ec44bc3faR634-R643

Reply:
I am not familiar enough with the topic but overall my advice to you is to, unless if obviously inefficient and wrong, optimize for readability and maintenance. Micro optimizations at the FRAME storage level at the cost of hard to maintain code are IMO a mistake. 

In this case, the amount of time we spent to go to the storage is way more significant than the length of the vector, so I reside to the above advice, do whatever is more sensible as if there is no weight/resource restriction.